### PR TITLE
[RFC] Support for fetching summarized descriptor data

### PR DESCRIPTION
**Pull request by hammem**
We recently added support for fetching ‘summary’ data from the
ThreatIndicator -> ThreatDescriptor connection in the ThreatExchange
RESTful APIs.

The summary object will return rolled up data for the descriptors
attached to a ThreatIndicator.  For now, there’s two summary fields:
- count (the number of descriptors tied to the indicator)
- preferred (when there are multiple descriptors, the one that is
  ‘preferred’)

Why do this?  When you encounter an indicator with multiple
descriptors, you have to do your own munging to handle it gracefully.
Rather than everyone have do the work on their own, the APIs offer a
way to do it for you (if you want it to).

But how does the underlying ThreatExchange API determine the
‘preferred’ one?

Through a set of lightweight, pre-defined heuristics, which can be
selected from in the API request.  Initially, the following heuristics
are supported:
- MINE_OR_NEWEST: the preferred descriptor should be ‘mine’, if I’ve
  submitted one.  failing that, return the newest published descriptor;
- NEWEST:  return the newest published descriptor;
- OLDEST:  return the oldest published descriptor;
- FROM_LIST_OR_NEWEST: the preferred descriptor should be one
  published by the member IDs from a provided list (order matters), if
  they’ve submitted one.  failing that, return the newest published
  descriptor.

We can support more heuristics in time, but started simple from the
methods we’d heard from those rolling their own logic.

There will also be support for summary data on other connections (e.g.
malware to indicators, malware family to variants, etc), so the support
is embedded in the Common connection() code in Pytx.

### CODE REVIEW

File: pytx/pytx/common.py [hunk 1]
```diff
@@ -411,6 +413,8 @@ def connections(cls_or_self,
         :type verify: bool, str
         :param metadata: Get extra metadata in the response.
         :type metadata: bool
+        :param summary: Get a summarized view of the connection.
+        :type summary: bool
```
Comment (mgoffin) [inline pytx/pytx/common.py]: We will need to add `summary_param` below this so people will know what to pass in.

File: pytx/pytx/common.py [hunk 2]
```diff
@@ -429,6 +433,15 @@ def connections(cls_or_self,
             params[t.FIELDS] = ','.join(f.strip() for f in fields)
         if metadata:
             params[t.METADATA] = 1
+        if summary and fields is not None:
+            params[t.FIELDS] = 'summary'
+        if summary:
+            params[t.FIELDS] = params[t.FIELDS] + ',summary'
```
Comment (mgoffin) [inline pytx/pytx/common.py]: This conditional will always overwrite the one above it. Should the one above it be `if summary and fields is None` and turn this one into an `elif summary`?

**Comment by mgoffin** [comment]
Other than the few comments, without knowing what should be passed in as `summary_params` it looks ok. One question: would it be easier to just assume if the developer has set `summary_params` then they want a summary, otherwise they don't? It would save on the extra `summary` argument to `connections()`. Or use `summary` as the only argument and it contains the parameters being defined?

File: pytx/pytx/common.py (hunk 1 above)
Comment (hammem) [inline pytx/pytx/common.py]: will do

File: pytx/pytx/common.py (hunk 2 above)
Comment (hammem) [inline pytx/pytx/common.py]: yep, good catch.

**Comment by hammem** [comment]
This is, likely, the first summarized edge we'll support.  This one has params that you can provide, but others may not, which is why I kept the two.  It's clumsy for this example, is there a better way ppl usually handle this?

**Comment by mgoffin** [comment]
Interesting. I didn't think about that. I suspect there might be other ways but a "yes/no" with optional params seems legit?

**Comment by hammem** [comment]
Given that we don't have one yet, I'll yank it and wait until we actually need it.

**Comment by hammem** [comment]
Updated with @mgoffin's changes.

File: pytx/pytx/common.py [hunk 3]
```diff
@@ -411,6 +412,12 @@ def connections(cls_or_self,
         :type verify: bool, str
         :param metadata: Get extra metadata in the response.
         :type metadata: bool
+
+        :param summary_params: Parameters to indicate a summary of the
+                               connections is desired and how to compute the
+                               summary.  Use pytx.vocabulary.SummaryConnection
+                               for the available params.
+        :type summary_params: map
```
Comment (mgoffin) [inline pytx/pytx/common.py]: `map` should be `dict` here I believe?

**Comment by RyPeck** [comment]
A few heuristics suggestions -

- Highest confidence
- Highest review rating status type (manual review being the highest)
- Lowest or highest StatusType (based on a scale of NON_MALICIOUS, SUSPICIOUS and MALICIOUS).

**Comment by hammem** [comment]
Thanks, @RyPeck .  For the last two, does having ties get broken by timestamp suffice?  That is, if there are two MALICIOUS descriptors, we just pick the newer of the two?

**Comment by RyPeck** [comment]
@hammem that suffices. Sorry for the delayed response.

**Comment by Dcallies** [comment]
Hey all, looks like this has staled out. I'm going to close this out unless you want to try and revive it!

**Comment by Dcallies** [comment]
```
               ...
             ;::::;
           ;::::; :;
         ;:::::'   :;
        ;:::::;     ;.
       ,:::::'       ;           OOO\
       ::::::;       ;          OOOOO\
       ;:::::;       ;         OOOOOOOO
      ,;::::::;     ;'         / OOOOOOO
    ;:::::::::`. ,,,;.        /  / DOOOOOO
  .';:::::::::::::::::;,     /  /     DOOOO
 ,::::::;::::::;;;;::::;,   /  /        DOOO
;`::::::`'::::::;;;::::: ,#/  /          DOOO
:`:::::::`;::::::;;::: ;::#  /            DOOO
::`:::::::`;:::::::: ;::::# /              DOO
`:`:::::::`;:::::: ;::::::#/               DOO
 :::`:::::::`;; ;:::::::::##                OO
 ::::`:::::::`;::::::::;:::#                OO
 `:::::`::::::::::::;'`:;::#                O
  `:::::`::::::::;' /  / `:#
   ::::::`:::::;'  /  /   `#
```

**Pull request closed without merging**